cli/azd/pkg/project/framework_service_external.go

```diff
@@ -0,0 +1,531 @@
+		RequestId: uuid.NewString(),
+		MessageType: &azdext.FrameworkServiceMessage_BuildRequest{
+			BuildRequest: &azdext.FrameworkServiceBuildRequest{
+				ServiceConfig: protoServiceConfig,
+				RestoreOutput: protoRestoreOutput,
+			},
+		},
+	}
+
+	resp, err := efs.sendAndWaitWithProgress(ctx, req, progress, func(r *azdext.FrameworkServiceMessage) bool {
+		return r.GetBuildResponse() != nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	buildResp := resp.GetBuildResponse()
+	if buildResp == nil {
+		return nil, fmt.Errorf("received empty build response")
+	}
+
+	return efs.fromProtoServiceBuildResult(buildResp.BuildResult), nil
+}
+
+// Package packages the source suitable for deployment
+func (efs *ExternalFrameworkService) Package(
+	ctx context.Context,
+	serviceConfig *ServiceConfig,
+	buildOutput *ServiceBuildResult,
+	progress *async.Progress[ServiceProgress],
+) (*ServicePackageResult, error) {
+	cleanup := efs.wireConsole()
+	defer cleanup()
+
+	protoServiceConfig, err := efs.toProtoServiceConfig(serviceConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	protoBuildOutput := efs.toProtoServiceBuildResult(buildOutput)
+
+	req := &azdext.FrameworkServiceMessage{
+		RequestId: uuid.NewString(),
+		MessageType: &azdext.FrameworkServiceMessage_PackageRequest{
+			PackageRequest: &azdext.FrameworkServicePackageRequest{
+				ServiceConfig: protoServiceConfig,
+				BuildOutput:   protoBuildOutput,
+			},
+		},
+	}
+
+	resp, err := efs.sendAndWaitWithProgress(ctx, req, progress, func(r *azdext.FrameworkServiceMessage) bool {
+		return r.GetPackageResponse() != nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	packageResp := resp.GetPackageResponse()
+	if packageResp == nil {
+		return nil, fmt.Errorf("received empty package response")
+	}
+
+	return fromProtoServicePackageResult(packageResp.PackageResult, nil), nil
+}
+
+// Private methods for gRPC communication
+
+// helper to send a request and wait for the matching response using async dispatcher
+func (efs *ExternalFrameworkService) sendAndWait(
+	ctx context.Context,
+	req *azdext.FrameworkServiceMessage,
+	match func(*azdext.FrameworkServiceMessage) bool,
+) (*azdext.FrameworkServiceMessage, error) {
+	// Create a response channel for this request
+	respChan := make(chan *azdext.FrameworkServiceMessage, 1)
+	efs.responseChans.Store(req.RequestId, respChan)
+	defer efs.responseChans.Delete(req.RequestId)
+
+	// Send the request
+	if err := efs.stream.Send(req); err != nil {
+		return nil, fmt.Errorf("failed to send request: %w", err)
+	}
+
+	// Wait for response via the async dispatcher
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case resp, ok := <-respChan:
+		if !ok {
+			return nil, fmt.Errorf("response channel closed")
+		}
+
+		if resp.Error != nil {
+			return nil, fmt.Errorf("framework service error: %s", resp.Error.Message)
+		}
+
+		if !match(resp) {
+			return nil, fmt.Errorf("received unexpected response type")
+		}
+
+		return resp, nil
+	}
+}
```

Comment:
After we get this in we should revisit to see if we can put some of the funs into a helper/base class to generalize the solution for the grpc communication pattner.